graphix/sim/density_matrix.py

```diff
@@ -313,11 +315,30 @@ def apply_channel(self, channel: KrausChannel, qargs):
         if not np.allclose(self.rho.trace(), 1.0):
             raise ValueError("The output density matrix is not normalized, check the channel definition.")
 
+    def subs(self, variable, substitute) -> DensityMatrix:
+        """Return a copy of the density matrix where all occurrences
+        of the given variable in measurement angles are substituted by
+        the given value.
+
+        See https://github.com/TeamGraphix/graphix-symbolic for
+        a symbolic parameter implementation that supports simulation.
+        """
+        result = DensityMatrix(nqubit=self.Nqubit)
+        result.rho = np.vectorize(lambda value: graphix.parameter.subs(value, variable, substitute))(self.rho)
+        return result
+
 
 class DensityMatrixBackend(Backend):
     """MBQC simulator with density matrix method."""
 
-    def __init__(self, pattern, max_qubit_num=12, pr_calc=True, input_state: Data = graphix.states.BasicStates.PLUS):
+    def __init__(
+        self,
+        pattern,
+        max_qubit_num=12,
+        pr_calc=True,
+        input_state: Data = graphix.states.BasicStates.PLUS,
+        rng: np.random.Generator | None = None,
+    ):
```

Comment:
```suggestion
    ) -> None:
```

Reply:
Type annotations for the whole `density_matrix.py` module have been added in 173c70e.